ports/unix/mpbthciport.c

```diff
@@ -60,12 +61,23 @@ extern bool mp_bluetooth_hci_poll(void);
 // For synchronous mode, we run all BLE stack code inside a scheduled task.
 // This task is scheduled periodically (every 1ms) by a background thread.
 
+// A signal to use to signal the main thread that a scheduled function is ready.
+// TODO how to choose this signal value?
+#define SCHED_SIG (SIGRTMIN + 4)
+
 // Allows the stack to tell us that we should stop trying to schedule.
 extern bool mp_bluetooth_hci_active(void);
 
 // Prevent double-enqueuing of the scheduled task.
 STATIC volatile bool events_task_is_scheduled = false;
 
+// Parent thread.
+// TODO: what about multithreading?
```

Comment:
FWIW, I did something similar a while back: https://github.com/pybricks/micropython/compare/966897f15149292615d068aba388bfe91c45c9d8...pybricks:f2b7d97ed31b3d22216ea69c356401ef9275e28f

However, I was interested in interrupting threads to raise exceptions rather than using the scheduler. 

The solution I came up with was to do like CPython and ensure that the scheduler only runs on the main thread.

Reply:
Ok, interesting solution.  Running the scheduler only on the main thread does simplify things.

But I think my approach here with signals is still broken, it has race conditions (the signal may arrive just before calling a blocking syscall like socket read, and then the scheduled code will not be run until the blocking call finishes).

Reply:
I see what you mean. That wouldn't work so well.

Reply:
FWIW, the other day, I came across the [pselect()](https://www.man7.org/linux/man-pages/man2/select.2.html) syscall that handles this race condition. However, I wasn't able to find other system calls that do the same.